mkolbina/lab3/main.c

```diff
@@ -0,0 +1,148 @@
+    char tmp = *a;
+    *a = *b;
+    *b = tmp;
+}
+
+int GetDirNamePos(char* path, int pathLength) {
+    int dirNamePos = 0;
+    for (int i = pathLength - 1; i >= 0; --i) {
+        if (path[i] == '/') {
+            dirNamePos = i + 1;
+            break;
+        }
+    }
+    return dirNamePos;
+}
+
+char* GetPathReverseDir(char* path) {
+    int pathLength = strlen(path);
+    char* reverseDirName = (char*)malloc(pathLength);
+    strcpy(reverseDirName, path);
+
+    int dirNamePos = GetDirNamePos(path, pathLength);
+    for (int i = dirNamePos; i < (pathLength + dirNamePos) / 2; ++i) {
+        swap(&(reverseDirName[i]), &(reverseDirName[pathLength - 1 - i + dirNamePos]));
+    }
+
+    return reverseDirName;
+}
+
+int GetSeporatorPossition(char* str, int length, char seporator) {
+    int pos = 0;
+    for (pos; pos < length; ++pos) {
+        if (str[pos] == seporator) break;
+    }
+
+    return pos;
+}
+
+char* GetReverseFileName(char* fileName, int lenFileName) {
+    char* reverseFileName = (char*)malloc(lenFileName);
+    memcpy(reverseFileName, fileName, lenFileName);
+
+    int seporPos = GetSeporatorPossition(fileName, lenFileName, '.');
+    for (int i = 0; i < seporPos / 2; ++i) {
+        swap(&reverseFileName[i], &reverseFileName[seporPos - i - 1]);
+    }
+
+    return reverseFileName;
+}
+
+char* Concatenate(char* pathDir, int lenPathDir, char* fileName, int lenFileName) {
+    int resultLength = lenPathDir + 1 + lenFileName;
+    char* result = (char*)malloc(resultLength);
+    memcpy(result, pathDir, lenPathDir);
+    result[lenPathDir] = '/';
+    memcpy(result + lenPathDir + 1, fileName, lenFileName);
+
+    return result;
+}
+
+void MakeReverseFile(char* pathOrigDir, char* pathReverseDir, char* fileName) {
+    int lenFileName = 0;
+    for (int i = 0; fileName[i] != '\0'; ++i) {
+        lenFileName++;
+    }
+
+    FILE* origFile, * reverseFile;
+
+    char* pathOrigFile = Concatenate(pathOrigDir, strlen(pathOrigDir), fileName, lenFileName);
+    if ((origFile = fopen(pathOrigFile, "rb")) == NULL) {
+        printf("Error: unable to open this file:\n%s\n", pathOrigFile);
+        free(pathOrigFile);
+        return;
+    }
+
+    char* reverseFileName = GetReverseFileName(fileName, lenFileName);
+    char* pathReverseFile = Concatenate(pathReverseDir, strlen(pathReverseDir), reverseFileName, lenFileName);
+    if ((reverseFile = fopen(pathReverseFile, "wb")) == NULL) {
+        printf("Error: unable to open this file:\n%s\n", pathReverseFile);
+        free(pathOrigFile);
+        free(pathReverseFile);
+        fclose(origFile);
+        return;
+    }
+
+    fseek(origFile, -1, SEEK_END);
+    int sizeOrigFile = ftell(origFile);
+    for (int filePos = sizeOrigFile; filePos >= 0; --filePos) {
+        fseek(origFile, filePos, SEEK_SET);
+        fputc(fgetc(origFile), reverseFile);
+    }
+
+    free(pathOrigFile);
+    free(reverseFileName);
+    free(pathReverseFile);
+    fclose(origFile);
```

Comment:
fclose - системный вызов, который может сломаться по многим причинам (неправильный файловый дескриптор, во время выполнения fclose снаружи процессу прилетел сигнал, ошибка чтения/записи). Кроме того, 

> Dealing with error returns from close()
A  careful  programmer will check the return value of close(), since it is quite possible that errors on a previous write(2) operation are reported only on the final close() that releases the open file description.  Failing to check the return value when closing a file may lead to silent loss of data.  This can especially be observed with NFS and with disk quota.

Reply:
Общее правило - нужно проверять возвращаемое значение у всех системных вызовов